polaris/polaris/management/commands/watch_transactions.py

```diff
@@ -63,51 +62,100 @@ def process_withdrawal(response, transaction):
     if response_memo != format_memo_horizon(transaction.withdraw_memo):
         return False
 
-    horizon_tx = TransactionEnvelope.from_xdr(envelope_xdr, network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE).transaction
+    horizon_tx = TransactionEnvelope.from_xdr(
+        response["envelope_xdr"],
+        network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE
+    ).transaction
     found_matching_payment_op = False
     for operation in horizon_tx.operations:
         if _check_payment_op(operation, transaction.asset.code, transaction.amount_in):
             found_matching_payment_op = True
             break
 
-    # If no matching payment operation is found in the Stellar response, return.
-    if not found_matching_payment_op:
-        return False
+    return found_matching_payment_op
+
 
-    # If the Stellar transaction succeeded, we mark the corresponding `Transaction`
-    # accordingly in the database. Else, we mark it `pending_stellar`, so the wallet
-    # knows to resubmit.
-    if successful:
+def _check_payment_op(operation: Operation, want_asset: str, want_amount: float) -> bool:
+    return (operation.type_code() == Xdr.const.PAYMENT and
+            str(operation.destination) == settings.STELLAR_DISTRIBUTION_ACCOUNT_ADDRESS and
+            str(operation.asset.code) == want_asset and
+            # TODO: Handle multiple possible asset issuance accounts
+            str(operation.asset.issuer) == settings.STELLAR_ISSUER_ACCOUNT_ADDRESS and
+            float(operation.amount) == want_amount)
+
+
+def update_transaction(response: Dict,
+                       transaction: Transaction,
+                       error_msg: str = None):
+    """
+    Updates the transaction depending on whether or not the transaction was
+    successfully executed on the Stellar network and `process_withdrawal`
+    completed without raising an exception.
+
+    If the Horizon response indicates the response was not successful, we mark
+    the status as `pending_stellar`. If exception was raised during
+    `process_withdrawal`, we mark the corresponding `Transaction` status as
+    `error` and save an error message to the transaction database object.
+    If the Stellar transaction succeeded, we mark it as `completed`.
+
+    :param error_msg: a description of the error that has occurred.
+    :param response: a response body returned from Horizon for the transaction
+    :param transaction: a database model object representing the transaction
+    """
+    if not response["successful"]:
```

Comment:
What is the difference between `not successful` and having an error?  wouldn't someting with an error_msg also be not successful?

Reply:
I don't particularly like the logic here either, I think its confusing, but I wanted to maintain our current functionality. Previously, if `response["successful"]` was `False`, we would save the `Transaction.status` as `pending_stellar`. Note that we now save `error_msg` to `Transaction.status_message` in this case.

As @debnil explained it, that status means the transaction was submitted to the stellar network but not "confirmed" as per [SEP-24](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0024.md#fields-for-withdraw-transactions). If we choose to start assigning unsuccessful transactions the `error` status, there also isn't a distinction between Stellar network failures and anchor server errors.

Errors originating from processing the transaction after it was submitted should use the `error` status. Specifically, errors raised in `process_withdrawal()`.

I'm totally open to suggestions on how this should work. Maybe we should think harder about error handling and reporting outside of this PR, though. Right now we don't provide good support for debugging issues related to failed transactions.